# Merge with conditions

Using the raw data from the **product_prices_cleaned.csv** file, perform the task of finding out how many times a lower price has been quoted for a given product, province, month historically. To do this, perform the following steps:

1. merge the table with itself. What type of join should be used?
1. filter the data to find earlier years and the values smaller than the one from current year in a given province.
1. group the data accordingly.

Which product(s) had the most such occurrences?

> See what happens to the column names when the columns are not used as a merge condition, but have the same names

In [1]:
import pandas as pd

# Load raw data
prices_path = '../../05_Day 5 - Pandas/01_Data/product_prices_cleaned.csv'
df = pd.read_csv(prices_path, sep=';')

# Parse year and month
df['year'] = df['date'].apply(lambda x: int(x.split('-')[0]))
df['month'] = df['date'].apply(lambda x: int(x.split('-')[1]))

# 1. Merge the table with itself.
# What type of join should be used?
# We use an 'inner' join to compare prices for the same product, province, and month.
merged = df.merge(df, on=['product', 'province', 'month'], suffixes=('_curr', '_hist'))

# 2. Filter the data to find earlier years and values lower than that in a given year.
filtered = merged[(merged['year_hist'] < merged['year_curr']) & (merged['value_hist'] < merged['value_curr'])]

# 3. Group the data accordingly.
product_counts = filtered.groupby('product').size().sort_values(ascending=False)

print("Product counts of historically lower prices:")
print(product_counts)

# Which product(s) had the most such occurrences?
most_occurrences = product_counts.idxmax()
max_count = product_counts.max()
print(f"\nThe product with the most occurrences is: '{most_occurrences}' with {max_count} occurrences.")



Product counts of historically lower prices:
product
beef with bone (rump steak) - per 1kg               38417
barley groats sausage - per 1kg                     38382
boneless beef (sirloin) - per 1kg                   38126
Italian head cheese - per 1kg                       37737
pork  meat (raw bacon) - per 1kg                    37571
smoked bacon with ribs - per 1kg                    37304
pork ham cooked - per 1kg                           34932
pork belly cooked - per 1kg                         34577
plain mixed bread (wheat-rye) - per 1kg             34109
dressed chickens - per 1kg                          32229
pork  with bone (center-cut pork chop) - per 1kg    31315
Hunter's sausage dried - per 1kg                    31309
pork meat with bone (shoulder) - per 1kg            31009
salted herring, non-dressed - per 1kg               25250
whole pickled cucumbers 0.9l - per 1pc.             22225
fresh chicken eggs - per 10pcs.                     22015
frozen carrot and p

### Questions & Answers:

1. **What type of join should be used?**
   An **inner join** should be used. We want to align each record with its historical counterparts for the exact same product, province, and month, matching only records that exist on both sides.

2. **What happens to the column names when the columns are not used as a merge condition, but have the same names?**
   Pandas appends suffixes to identify them uniquely (by default `_x` and `_y`, or the custom `suffixes=('_curr', '_hist')` we specified in the `merge` call).

3. **Which product(s) had the most such occurrences?**
   The product with the most occurrences of historically lower prices is **'beef with bone (rump steak) - per 1kg'** with **38,417** occurrences.

